Let Jest handle all file types

gaearon opened the pull request:
In 0.8.x, we switched to treat all non-JS/CSS extensions as files by default (#1059, #1077).

However when we tried to do the same in Jest, we made a mistake which introduced false positives (#1145, #1147).

We temporarily reverted this change in #1149 and reopened https://github.com/facebookincubator/create-react-app/issues/667#issuecomment-264693934. @cpojer suggested to use `transform` instead of `moduleNameMapper` so that we can match against the full paths.

This PR implements that. I verified that the following cases work:

* Importing CSS both inside project directory and in `node_modules` (gives you `{})
* Importing arbitrary files both inside project directory and in `node_modules` (gives you the filename)
* Importing JS modules like `../something` or `lodash.assign` (covers #1145, #1147).

Fixes #667.

gaearon commented on packages/react-scripts/utils/createJestConfig.js:
```diff
@@ -19,25 +19,28 @@ module.exports = (resolve, rootDir, isEjecting) => {
+      '^.+\\.(?!(js|jsx|css|json)$)[^\\.]+$': resolve('config/jest/fileTransform.js'),
+    },
+    transformIgnorePatterns: [
+      '/node_modules/.+\.(js|jsx|json)$'
```
I needed this because otherwise we can't import CSS from `node_modules` (e.g. `bootstrap/dist/bootstrap.css`). Would it make sense to change the default to that?

cpojer commented on packages/react-scripts/config/jest/fileTransform.js:
```diff
@@ -5,9 +5,16 @@
+
+module.exports = {
+  process(src, filename) {
+    return 'module.exports = ' + JSON.stringify(path.basename(filename)) + ';';
```
should this not include the extension name?

gaearon replied:
I think it does!

```js
 require('path').basename('aas.asa')
'aas.asa'

```

cpojer commented:
nice!

gaearon commented:
Tested with:

```js
import React, { Component } from 'react';
import appCss from './App.css';
import depCss from 'postcss-discard-overridden/test/input.css';
import appSVG from './logo.svg';
import depSVG from 'postcss-load-options/logo.svg';
import appNoExt from './noext';
import depNoExt from 'abbrev/LICENSE';
import appUnknown from './lol.unknown';  // I added this
import depUnknown from 'abab/lol.unknown'; // I added this
import depJS from 'abab/lol'; // I added this, it's just module.exports = function() { return `haha` }

class App extends Component {
  render() {
    return (
      <div className="App">
        app css: {JSON.stringify(appCss)}
        <br />{'\n'}
        dep css: {JSON.stringify(depCss)}
        <br />{'\n'}
        app svg: {appSVG}
        <br />{'\n'}
        dep svg: {depSVG}
        <br />{'\n'}
        app noExt: {appNoExt}
        <br />{'\n'}
        dep noExt: {depNoExt}
        <br />{'\n'}
        app unknown: {appUnknown}
        <br />{'\n'}
        dep unknown: {depUnknown}
        <br />{'\n'}
        dep js {depJS()}{depJS.toString()}
      </div>
    );
  }
}

export default App;
```

(Need to add something like this to integration test, see #1187)

`npm start`

<img width="505" alt="screen shot 2016-12-07 at 20 13 12" src="https://cloud.githubusercontent.com/assets/810438/20984901/175b74aa-bcba-11e6-8091-8bb7dfa6f8b4.png">

`npm test`

<img width="346" alt="screen shot 2016-12-07 at 20 13 37" src="https://cloud.githubusercontent.com/assets/810438/20984907/1aa57cd2-bcba-11e6-904e-74c6639c3200.png">

`npm start` post ejection

<img width="393" alt="screen shot 2016-12-07 at 20 15 19" src="https://cloud.githubusercontent.com/assets/810438/20984918/20739b8a-bcba-11e6-8ee3-bccc88000907.png">

`npm test` post ejection

<img width="361" alt="screen shot 2016-12-07 at 20 15 25" src="https://cloud.githubusercontent.com/assets/810438/20984926/26192730-bcba-11e6-8d92-4007b8d23430.png">

EnoahNetzach commented:
I already added this to the todo list of #1187 ;)
